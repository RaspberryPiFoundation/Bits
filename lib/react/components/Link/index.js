var _jsxFileName = 'src/components/Link/index.js',
    _this = this;

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { Link as RouterLink } from 'react-router-dom';
// import './Link.scss'

var Link = function Link(_ref) {
  var className = _ref.className,
      to = _ref.to,
      props = _objectWithoutProperties(_ref, ['className', 'to']);

  var classes = classnames('c-link', className);

  return React.createElement(
    RouterLink,
    Object.assign({ className: classes, to: to }, props, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 11
      },
      __self: _this
    }),
    children
  );
};

Link.propTypes = {
  className: PropTypes.string,
  to: PropTypes.string.isRequired
};

export default Link;