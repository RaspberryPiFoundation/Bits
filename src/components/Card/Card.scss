@import '../../styles/component-helper';

$card-background-color: $color-white;
$card-border-radius:    $border-radius;
$card-box-shadow:       $box-shadow;
$card-content-p:        $baseline-x4;
$card-background-color: $color-white;

.c-card {
  background-color: $card-background-color;
  border-radius:    $card-border-radius;
  box-shadow:       $card-box-shadow;
  overflow:         hidden;
}

.c-card__content {
  padding: $card-content-p;
}

.c-card__title {
  @include type-sub-heading;
}
