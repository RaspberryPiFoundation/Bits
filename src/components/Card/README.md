# Card

Cards are used a lot, and are fairly flexible. The contents of a card should never be too complicated. An optional image can be added, and the whole card can be a link if necessary (or you can add a regular link to the card's content too if you prefer). They will always stretch to 100% the width of their parent, and are intended for use within the iota grid system.

## With Image

```jsx
<Card
  imageSrc="http://via.placeholder.com/300x200"
  imageAlt="Placeholder image">
    <p class="o-type-base u-mb-base">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
    <span class="o-type-caption">2 days ago</span>
</Card>
```

#### In HTML/CSS

```html static
<div class="c-card">
  <img src="http://via.placeholder.com/300x200" alt="Placeholder image" class="c-card__image" />
  <div class="c-card__content">
    <p class="o-type-base u-mb-base">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
    <span class="o-type-caption">2 days ago</span>
  </div>
</div>
```

## Without Image

```jsx
<Card>
  <p class="o-type-base u-mb-base">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
  <span class="o-type-caption">2 days ago</span>
</Card>
```

#### In HTML/CSS

```html static
<div class="c-card" href="#">
  <div class="c-card__content">
    <p class="o-type-base u-mb-base">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
    <span class="o-type-caption">2 days ago</span>
  </div>
</div>
```

## As Link

```jsx
import { MemoryRouter } from 'react-router-dom';
<MemoryRouter>
  <Card
    imageSrc="http://via.placeholder.com/300x200"
    imageAlt="Placeholder image"
    to="https://www.raspberrypi.org">
      <p class="o-type-base u-mb-base">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
      <span class="o-type-caption">2 days ago</span>
  </Card>
</MemoryRouter>
```

#### In HTML/CSS

```html static
<a class="c-card c-card--link" href="#">
  <img src="http://via.placeholder.com/300x200" alt="" class="c-card__image" />
  <div class="c-card__content">
    <p class="o-type-base u-mb-base">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
    <span class="o-type-caption">2 days ago</span>
  </div>
</a>
```
