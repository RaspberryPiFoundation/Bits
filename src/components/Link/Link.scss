@import '../../styles/component-helper';

.c-link {
  @include link;
}

.c-link--light-on-dark {
  @include link-light-on-dark;
}
