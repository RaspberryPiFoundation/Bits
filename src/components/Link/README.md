```jsx
import { MemoryRouter } from 'react-router-dom';
<MemoryRouter>
  <Link to="https://www.raspberrypi.org">Link text goes here</Link>
</MemoryRouter>
```

#### In HTML/CSS

```html static
<a class="c-link" href="https://www.raspberrypi.org">Link text goes here</a>
```
