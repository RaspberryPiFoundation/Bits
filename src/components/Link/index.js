import { Base } from 'react-iotacss'
import { Link as RouterLink } from 'react-router-dom'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import './Link.scss'

const Link = ({ children, className, excludeClassname, to, ...props }) => {
  const classNames = classnames(
    {
      'c-link': !excludeClassname,
    },
    className,
  )

  const anchorLink = (
    <Base className={classNames} href={to} tagName="a" {...props}>
      {children}
    </Base>
  )

  const isExternal = () => !/^https?:\/\//i.test(to)

  const routerLink = (
    <RouterLink className={classNames} to={to} {...props}>
      {children}
    </RouterLink>
  )

  if (isExternal()) {
    return anchorLink
  }

  return routerLink
}

Link.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  excludeClassname: PropTypes.bool,
  to: PropTypes.string.isRequired,
}

export default Link
