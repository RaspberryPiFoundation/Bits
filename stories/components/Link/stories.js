import React from 'react'
import { storiesOf } from '@storybook/react'

import { StaticRouter } from 'react-router-dom'

import Link from '../../../src/components/Link'

const context = {}

storiesOf('Link', module)
  .add('Internal link', () => (
    <StaticRouter context={context}>
      <Link to="/here">An internal link</Link>
    </StaticRouter>
  ))
  .add('Light on dark internal link', () => (
    <StaticRouter context={context}>
      <Link className="c-link--light-on-dark" to="/">
        Hello Link
      </Link>
    </StaticRouter>
  ))
  .add('external link', () => (
    <StaticRouter context={context}>
      <Link to="https://raspberrypi.org">An external link</Link>
    </StaticRouter>
  ))
  .add('Light on dark external link', () => (
    <StaticRouter context={context}>
      <Link className="c-link--light-on-dark" to="https://raspberrypi.org">
        Hello Link
      </Link>
    </StaticRouter>
  ))
