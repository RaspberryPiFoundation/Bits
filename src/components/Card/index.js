import { Base } from 'react-iotacss'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import './Card.scss'
import Link from '../Link'

const Card = ({
  children,
  className,
  heading,
  imageAlt,
  imageSrc,
  imageTitle,
  to,
  ...props
}) => {
  const isLink = () => typeof to === 'string'

  const classNames = classnames('c-card', className, {
    'c-card--link': isLink(),
  })

  const imageJSX = () => {
    if (!imageSrc) {
      return null
    }

    return (
      <Base
        alt={imageAlt}
        className="c-card__image"
        src={imageSrc}
        tagName="img"
        title={imageTitle}
      />
    )
  }

  const headingJSX = () => {
    if (!heading) {
      return null
    }

    return <span className="c-card__heading">{heading}</span>
  }

  const cardChildren = () => (
    <div>
      {imageJSX()}
      <div className="c-card__content">
        {headingJSX()}
        {children}
      </div>
    </div>
  )

  if (isLink()) {
    return (
      <Link className={classNames} excludeClassname={true} to={to} {...props}>
        {cardChildren()}
      </Link>
    )
  }

  return (
    <Base className={classNames} {...props}>
      {cardChildren()}
    </Base>
  )
}

Card.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  className: PropTypes.string,
  heading: PropTypes.string,
  imageAlt: PropTypes.string,
  imageSrc: PropTypes.string,
  imageTitle: PropTypes.string,
  /** relative or qualified URL - determines if component renders `<a>` or `<div>` */
  to: PropTypes.string,
}

export default Card
