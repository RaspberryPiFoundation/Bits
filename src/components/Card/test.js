import { mount, shallow } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import Card from '.'
import Link from '../Link'
import React from 'react'

let mountedComp
let shallowComp

const children = <span>Whatever</span>
const heading = 'Card Heading'
const imageSrc = 'http://via.placeholder.com/350x150'
const to = 'https://raspberrypi.org'

describe('<Card />', () => {
  beforeEach(() => {
    shallowComp = shallow(
      <Card heading={heading} imageSrc={imageSrc}>
        {children}
      </Card>,
    )
  })

  describe('Renders', () => {
    beforeEach(() => {
      mountedComp = mount(
        <Card heading={heading} imageSrc={imageSrc}>
          {children}
        </Card>,
      )
    })

    it('without crashing', () => {
      expect(mountedComp.find('.c-card')).toBePresent()
    })

    it('matches its snapshot', () => {
      expect(mountedComp).toMatchSnapshot()
    })
  })

  describe('Props', () => {
    it('renders an image with the correct src', () => {
      expect(shallowComp.find('.c-card__image')).toHaveProp('src', imageSrc)
    })

    it('renders the title', () => {
      expect(shallowComp.find('.c-card__heading')).toIncludeText(heading)
    })

    it('renders the content', () => {
      expect(shallowComp.find('.c-card__content')).toHaveHTML(
        [
          '<div class="c-card__content">',
          '<span class="c-card__heading">',
          'Card Heading',
          '</span>',
          '<span>',
          'Whatever',
          '</span>',
          '</div>',
        ].join(''),
      )
    })

    it('renders the technologies element', () => {
      expect(shallowComp.find('span')).toBePresent()
    })
  })

  describe('Methods', () => {
    describe('#imageJSX', () => {
      it('returns null if no imageSrc prop', () => {
        shallowComp = shallow(<Card>{children}</Card>)
        console.log(shallowComp)
        let value = shallowComp.instance().imageJSX()
        expect(value).toEqual(null)
      })

      fit('returns HTML if imageSrc prop supplied', () => {
        shallowComp = shallow(<Card imageSrc={imageSrc}>{children}</Card>)
        let value = shallowComp.instance().imageJSX()
        expect(value).toEqual(null)
      })
    })
  })

  describe('Variants', () => {
    describe('...as Link', () => {
      beforeEach(() => {
        mountedComp = mount(
          <MemoryRouter>
            <Card heading={heading} imageSrc={imageSrc} to={to}>
              {children}
            </Card>
          </MemoryRouter>,
        )
      })

      it('links to the right place', () => {
        expect(mountedComp.find(Link)).toHaveProp('to', to)
      })
    })
  })
})
