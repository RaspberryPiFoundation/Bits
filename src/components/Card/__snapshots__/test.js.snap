// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Renders matches its snapshot 1`] = `
<Card
  heading="Card Heading"
  imageSrc="http://via.placeholder.com/350x150"
>
  <Base
    className="c-card"
    tagName="div"
  >
    <div
      className="c-card "
    >
      <div>
        <Base
          className="c-card__image"
          src="http://via.placeholder.com/350x150"
          tagName="img"
        >
          <img
            className="c-card__image "
            src="http://via.placeholder.com/350x150"
          />
        </Base>
        <div
          className="c-card__content"
        >
          <span
            className="c-card__heading"
          >
            Card Heading
          </span>
          <span>
            Whatever
          </span>
        </div>
      </div>
    </div>
  </Base>
</Card>
`;
